mindmeld/system_entity_recognizer.py

```diff
@@ -41,61 +44,101 @@ class SystemEntityError(Exception):
     pass
 
 
-class SystemEntityRecognizer:
+class SystemEntityRecognizer(ABC):
     """SystemEntityRecognizer is the external parsing service used to extract
     system entities. It is intended to be used as a singleton, so it's
     initialized only once during NLP object construction.
+    """
+
+    @staticmethod
+    def get_instance(app_path):
+        """ Static access method.
+        In general we want to find the system entity recognizer from the application config.
+        If the application configuration is empty, we do not use Duckling.
+        Otherwise, we return the Duckling recognizer with the URL defined in the application's
+          config, default to the DEFAULT_DUCKLING_URL.
+
+        Args:
+            app_path: The application path
+
+        Returns:
+            (SystemEntityRecognizer): A SystemEntityRecognizer instance
+        """
+        if is_duckling_configured(app_path):
+            url = get_system_entity_url_config(app_path=app_path)
+            return DucklingRecognizer.get_instance(url)
+        else:
+            return DummySystemEntityRecognizer.get_instance()
+
+    @abstractmethod
+    def get_response(self, data):
+        pass
+
 
-    TODO: Abstract this class into an interface and implement the duckling
-    service as one such service.
+class DummySystemEntityRecognizer(SystemEntityRecognizer):
+    """
+    This is a dummy recognizer which returns empty list and NO_RESPONSE_CODE.
     """
 
     _instance = None
 
-    def __init__(self, app_path=None):
+    def __init__(self):
+        if not DummySystemEntityRecognizer._instance:
+            DummySystemEntityRecognizer._instance = self
+        else:
+            raise Exception("DummySystemEntityRecognizer is a singleton")
+
+    @staticmethod
+    def get_instance():
+        if not DummySystemEntityRecognizer._instance:
+            DummySystemEntityRecognizer()
+
+        return DummySystemEntityRecognizer._instance
+
+    def get_response(self, data):
+        del data
+        return [], NO_RESPONSE_CODE
+
+
+class DucklingRecognizer(SystemEntityRecognizer):
+    _instances = {}
```

Comment:
why have this dict?

Reply:
I am thinking of having a singleton instance per URL. What do you think? We can change to a singleton for all URL as well.

Reply:
isnt there only one URL per `SystemEntityRecognizer ` class? If that is the case, changing it to string or single object makes sense.

Reply:
Can we remove this?

Reply:
Ok